File: user-service/src/clients/event.client.js
const grpc = require("@grpc/grpc-js");
const protoLoader = require("@grpc/proto-loader");
const path = require("path");

class EventServiceClient {
  constructor(serverAddress = "event-service:50055") {
    // Load the protobuf definition
    const PROTO_PATH = path.join(__dirname, "../protos/event.proto");
    const packageDefinition = protoLoader.loadSync(PROTO_PATH, {
      keepCase: true,
      longs: String,
      enums: String,
      defaults: true,
      oneofs: true,
    });

    const eventProto = grpc.loadPackageDefinition(packageDefinition).event;

    // Create the gRPC client
    this.client = new eventProto.EventService(
      process.env.EVENT_SERVICE_HOST || serverAddress,
      grpc.credentials.createInsecure()
    );
  }

  /**
   * Publish a domain event
   */
  async publishEvent({
    aggregateId,
    aggregateType,
    eventType,
    eventData,
    metadata = {},
    correlationId = "",
    causationId = "",
  }) {
    return new Promise((resolve, reject) => {
      const request = {
        aggregate_id: aggregateId,
        aggregate_type: aggregateType,
        event_type: eventType,
        event_data: JSON.stringify(eventData),
        metadata: JSON.stringify(metadata),
        correlation_id: correlationId,
        causation_id: causationId,
      };

      this.client.publishEvent(request, (error, response) => {
        if (error) {
          reject(error);
        } else if (!response.success) {
          reject(new Error(response.message));
        } else {
          resolve(response);
        }
      });
    });
  }

  /**
   * Get events by type or recent events
   */
  async getEvents({
    eventType = "",
    aggregateType = "",
    limit = 100,
    offset = 0,
  } = {}) {
    return new Promise((resolve, reject) => {
      const request = {
        event_type: eventType,
        aggregate_type: aggregateType,
        limit,
        offset,
      };

      this.client.getEvents(request, (error, response) => {
        if (error) {
          reject(error);
        } else if (!response.success) {
          reject(new Error(response.message));
        } else {
          resolve(response.events);
        }
      });
    });
  }

  /**
   * Get events for a specific aggregate
   */
  async getEventsByAggregate({ aggregateId, aggregateType, fromVersion = 1 }) {
    return new Promise((resolve, reject) => {
      const request = {
        aggregate_id: aggregateId,
        aggregate_type: aggregateType,
        from_version: fromVersion,
      };

      this.client.getEventsByAggregate(request, (error, response) => {
        if (error) {
          reject(error);
        } else if (!response.success) {
          reject(new Error(response.message));
        } else {
          resolve(response.events);
        }
      });
    });
  }

  /**
   * Subscribe to events
   */
  async subscribeToEvents({ consumerGroup, eventTypes = [], callbackUrl }) {
    return new Promise((resolve, reject) => {
      const request = {
        consumer_group: consumerGroup,
        event_types: eventTypes,
        callback_url: callbackUrl,
      };

      this.client.subscribeToEvents(request, (error, response) => {
        if (error) {
          reject(error);
        } else if (!response.success) {
          reject(new Error(response.message));
        } else {
          resolve(response);
        }
      });
    });
  }

  /**
   * Close the client connection
   */
  close() {
    this.client.close();
  }
}

module.exports = EventServiceClient;

File: post-service/clients/event_client/event_client.go
package clients

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"

	pb "post-service/protos/eventpb"
)

// EventServiceClient wraps the gRPC client for event service
type EventServiceClient struct {
	client pb.EventServiceClient
	conn   *grpc.ClientConn
}

// NewEventServiceClient creates a new event service client
func NewEventServiceClient(serverAddress string) (*EventServiceClient, error) {
	if serverAddress == "" {
		serverAddress = "event-service:50055"
	}

	// Create connection with timeout
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	conn, err := grpc.DialContext(ctx, serverAddress,
		grpc.WithTransportCredentials(insecure.NewCredentials()),
		grpc.WithBlock(),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to connect to event service at %s: %w", serverAddress, err)
	}

	client := pb.NewEventServiceClient(conn)

	return &EventServiceClient{
		client: client,
		conn:   conn,
	}, nil
}

// PublishEvent publishes a domain event
func (c *EventServiceClient) PublishEvent(ctx context.Context, req *pb.PublishEventRequest) (*pb.PublishEventResponse, error) {
	// Marshal event data to JSON
	eventDataJSON, err := json.Marshal(req.EventData)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal event data: %w", err)
	}

	// Marshal metadata to JSON
	metadataJSON, err := json.Marshal(req.Metadata)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal metadata: %w", err)
	}

	// Create gRPC request
	grpcReq := &pb.PublishEventRequest{
		AggregateId:   req.AggregateId,
		AggregateType: req.AggregateType,
		EventType:     req.EventType,
		EventData:     string(eventDataJSON),
		Metadata:      string(metadataJSON),
		CorrelationId: req.CorrelationId,
		CausationId:   req.CausationId,
	}

	// Call the service
	response, err := c.client.PublishEvent(ctx, grpcReq)
	if err != nil {
		return nil, fmt.Errorf("failed to publish event: %w", err)
	}

	if !response.Success {
		return nil, fmt.Errorf("event publishing failed: %s", response.Message)
	}

	return response, nil
}

// GetEvents retrieves events by type or recent events
func (c *EventServiceClient) GetEvents(ctx context.Context, req *pb.GetEventsRequest) ([]*pb.Event, error) {
	// Set defaults
	if req.Limit == 0 {
		req.Limit = 100
	}

	grpcReq := &pb.GetEventsRequest{
		EventType:     req.EventType,
		AggregateType: req.AggregateType,
		Limit:         req.Limit,
		Offset:        req.Offset,
	}

	response, err := c.client.GetEvents(ctx, grpcReq)
	if err != nil {
		return nil, fmt.Errorf("failed to get events: %w", err)
	}

	if !response.Success {
		return nil, fmt.Errorf("get events failed: %s", response.Message)
	}

	return response.Events, nil
}

// GetEventsByAggregate retrieves events for a specific aggregate
func (c *EventServiceClient) GetEventsByAggregate(ctx context.Context, req *pb.GetEventsByAggregateRequest) ([]*pb.Event, error) {
	// Set default
	if req.FromVersion == 0 {
		req.FromVersion = 1
	}

	grpcReq := &pb.GetEventsByAggregateRequest{
		AggregateId:   req.AggregateId,
		AggregateType: req.AggregateType,
		FromVersion:   req.FromVersion,
	}

	response, err := c.client.GetEventsByAggregate(ctx, grpcReq)
	if err != nil {
		return nil, fmt.Errorf("failed to get events by aggregate: %w", err)
	}

	if !response.Success {
		return nil, fmt.Errorf("get events by aggregate failed: %s", response.Message)
	}

	return response.Events, nil
}

// SubscribeToEvents subscribes to events
func (c *EventServiceClient) SubscribeToEvents(ctx context.Context, req *pb.SubscribeToEventsRequest) (*pb.SubscribeToEventsResponse, error) {
	grpcReq := &pb.SubscribeToEventsRequest{
		ConsumerGroup: req.ConsumerGroup,
		EventTypes:    req.EventTypes,
		CallbackUrl:   req.CallbackUrl,
	}

	response, err := c.client.SubscribeToEvents(ctx, grpcReq)
	if err != nil {
		return nil, fmt.Errorf("failed to subscribe to events: %w", err)
	}

	if !response.Success {
		return nil, fmt.Errorf("subscription failed: %s", response.Message)
	}

	return response, nil
}

// Close closes the client connection
func (c *EventServiceClient) Close() error {
	if c.conn != nil {
		return c.conn.Close()
	}
	return nil
}
